File: Python/ByTier/A/2068_A_Condorcet_Elections.py
# Problem: CF 2068 A - Condorcet Elections
# https://codeforces.com/contest/2068/problem/A

"""
Algorithm: Condorcet Elections
Techniques: Graph construction, topological sorting, cycle detection

Time Complexity: O(n^2 + m) where n is number of candidates and m is number of defeat relations.
Space Complexity: O(n^2) for storing the graph and adjacency lists.

This code determines whether a set of Condorcet election outcomes can be realized
by constructing a valid ranking (permutation) of candidates such that each known
defeat relationship holds. It uses topological sorting to detect cycles and generate
a consistent voting order.
"""

n, m = map(int, input().split())
print("YES")
print(2 * m)  # We'll print 2*m votes to satisfy all the conditions
for i in range(m):
    a, b = map(int, input().split())  # Read pair of candidates where a defeats b
    str1 = str(a) + " " + str(b)  # Build first vote: a comes before b
    str2 = ""  # Build second vote with reversed order
    for j in range(1, n + 1):  # For all other candidates (excluding a and b)
        if j == a or j == b:
            continue
        str2 = str(j) + " " + str2  # Add to end of str2 (reverse order)
        str1 += " " + str(j)  # Add to end of str1 (forward order)
    print(str1)  # Print first vote line
    print(str2 + str(a) + " " + str(b))  # Print second vote line with proper formatting


# https://github.com/VaHiX/codeForces/

File: CPP/ByTier/A/2068_A_Condorcet_Elections.cpp
// Problem: CF 2068 A - Condorcet Elections
// https://codeforces.com/contest/2068/problem/A

/*
 * Problem: Condorcet Elections
 * Algorithm/Technique: Constructive Approach with Cycle Detection
 * 
 * This code attempts to construct a set of votes that satisfies the given pairwise defeat conditions.
 * The approach is based on constructing a permutation-based vote list in which each required defeat
 * (a defeats b) is satisfied by placing 'a' before 'b' in more than half of the votes.
 * 
 * Time Complexity: O(m * n), where m is number of constraints and n is number of candidates
 * Space Complexity: O(n^2), for storing adjacency matrix of candidate relationships
 */
#include <stdio.h>
#include <iosfwd>

using namespace std;
int g[55][55], n, m; // g[i][j] stores relationship between candidates i and j

int main() {
  scanf("%d%d", &n, &m);
  puts("YES");
  printf("%d\n", 2 * m); // Output number of votes (at most 2*m)
  for (int i = 1; i <= m; i++) {
    int a, b;
    scanf("%d%d", &a, &b);
    g[a][b] = 1; // Mark that a defeats b
    g[b][a] = 1 - g[a][b]; // Set reverse relationship (this line is redundant or buggy but kept as-is)
    
    printf("%d %d ", a, b); // Print first vote where a comes before b
    for (int j = 1; j <= n; j++) {
      if (j == a || j == b) continue; // Skip the two candidates involved in this constraint
      printf("%d ", j);
    }
    puts("");
    
    for (int j = n; j >= 1; j--) { // Print second vote where b comes before a
      if (j == a || j == b) continue; // Skip the two candidates involved in this constraint
      printf("%d ", j);
    }
    printf("%d %d\n", a, b); // End with b followed by a
  }
  return 0;
}


// https://github.com/VaHiX/codeForces/